base/server/src/com/netscape/cms/profile/common/CAEnrollProfile.java

```diff
@@ -105,12 +117,103 @@ public void execute(IRequest request)
             throw new EProfileException("No CA Service");
         }
 
+        // if isServerSideKeygen, send keygen request to KRA
+        boolean isSSKeygen = false;
+        String isSSKeygenStr = request.getExtDataInString("isServerSideKeygen");
+        if (isSSKeygenStr != null && isSSKeygenStr.equalsIgnoreCase("true")) {
+            logger.debug(method + "isServerSideKeygen = true");
+            isSSKeygen = true;
+        } else {
+            logger.debug(method + "isServerSideKeygen = false");
+        }
+
+        // prepare for auditing
+        CertificateSubjectName reqSubj =
+                request.getExtDataInCertSubjectName(IRequest.REQUEST_SUBJECT_NAME);
+        String clientId = "unknown serverKeyGenUser";
+        if (reqSubj != null) {
+            X500Name xN = reqSubj.getX500Name();
+            clientId = xN.toString();
+            logger.debug(method + "clientId = " + clientId);
+        }
+
         // if PKI Archive Option present, send this request
         // to DRM
         byte optionsData[] = request.getExtDataInByteArray(IRequest.REQUEST_ARCHIVE_OPTIONS);
 
-        // do not archive keys for renewal requests
-        if ((optionsData != null) && (!request.getRequestType().equals(IRequest.RENEWAL_REQUEST))) {
+        byte[] transWrappedSessionKey = null;
+        byte[] sessionWrappedPassphrase = null;
+        if (isSSKeygen) { // Server-Side Keygen enrollment
+            request.setExtData(IRequest.SSK_STAGE, IRequest.SSK_STAGE_KEYGEN);
+
+            /*
+             * temporarily remove the items not needed for SSK_STAGE_KEYGEN
+             * so not to pass them to KRA.
+             * They will be put back at SSK_STAGE_KEY_RETRIEVE below
+             */
+            transWrappedSessionKey = (byte[]) request.getExtDataInByteArray("serverSideKeygenP12PasswdTransSession");
+
+            sessionWrappedPassphrase = (byte[]) request.getExtDataInByteArray("serverSideKeygenP12PasswdEnc");
+
+            request.setExtData("serverSideKeygenP12PasswdTransSession", "");
+            request.deleteExtData("serverSideKeygenP12PasswdTransSession");
+            request.setExtData("serverSideKeygenP12PasswdEnc", "");
+            request.deleteExtData("serverSideKeygenP12PasswdEnc");
+
+            try {
+                IConnector kraConnector = caService.getKRAConnector();
+
+                if (kraConnector == null) {
```

Comment:
Should this be throwing an exception when the connector is null?

Reply:
will do.